plugins/outputs/wavefront/wavefront.go

```diff
@@ -0,0 +1,299 @@
+	if w.ConvertPaths && w.MetricSeparator == "_" {
+		w.ConvertPaths = false
+	}
+	if w.ConvertPaths {
+		pathReplacer = strings.NewReplacer("_", w.MetricSeparator)
+	}
+
+	// Test Connection to Wavefront proxy Server
+	uri := fmt.Sprintf("%s:%d", w.Host, w.Port)
+	tcpAddr, err := net.ResolveTCPAddr("tcp", uri)
+	if err != nil {
+		return fmt.Errorf("Wavefront: TCP address cannot be resolved %s", err.Error())
+	}
+	connection, err := net.DialTCP("tcp", nil, tcpAddr)
+	if err != nil {
+		return fmt.Errorf("Wavefront: TCP connect fail %s", err.Error())
+	}
+	defer connection.Close()
+	return nil
+}
+
+func (w *Wavefront) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+
+	// Send Data to Wavefront proxy Server
+	uri := fmt.Sprintf("%s:%d", w.Host, w.Port)
+	tcpAddr, _ := net.ResolveTCPAddr("tcp", uri)
+	connection, err := net.DialTCP("tcp", nil, tcpAddr)
+	if err != nil {
+		return fmt.Errorf("Wavefront: TCP connect fail %s", err.Error())
+	}
+	defer connection.Close()
+
+	for _, m := range metrics {
+		for _, metric := range buildMetrics(m, w) {
+			messageLine := fmt.Sprintf("%s %s %v %s\n", metric.Metric, metric.Value, metric.Timestamp, metric.Tags)
+			log.Printf("D! Output [wavefront] %s", messageLine)
+			_, err := connection.Write([]byte(messageLine))
+			if err != nil {
+				return fmt.Errorf("Wavefront: TCP writing error %s", err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func buildTags(mTags map[string]string, w *Wavefront) []string {
+	sourceTagFound := false
+
+	for _, s := range w.SourceOverride {
+		for k, v := range mTags {
+			if k == s {
+				mTags["source"] = v
+				mTags["telegraf_host"] = mTags["host"]
+				sourceTagFound = true
+				delete(mTags, k)
+				break
+			}
+		}
+		if sourceTagFound {
+			break
+		}
+	}
+
+	if !sourceTagFound {
+		mTags["source"] = mTags["host"]
+	}
+	delete(mTags, "host")
+
+	tags := make([]string, len(mTags))
+	index := 0
+	for k, v := range mTags {
+		if w.UseRegex {
+			tags[index] = fmt.Sprintf("%s=\"%s\"", sanitizedRegex.ReplaceAllString(k, "-"), tagValueReplacer.Replace(v))
+		} else {
+			tags[index] = fmt.Sprintf("%s=\"%s\"", sanitizedChars.Replace(k), tagValueReplacer.Replace(v))
+		}
+
+		index++
+	}
+
+	sort.Strings(tags)
+	return tags
+}
+
+func buildMetrics(m telegraf.Metric, w *Wavefront) []*MetricLine {
+	if w.DebugAll {
+		log.Printf("D! Output [wavefront] original name: %s\n", m.Name())
+	}
+
+	ret := []*MetricLine{}
+	for fieldName, value := range m.Fields() {
+		if w.DebugAll {
+			log.Printf("D! Output [wavefront] original field: %s\n", fieldName)
+		}
+
+		var name string
+		if !w.SimpleFields && fieldName == "value" {
+			name = fmt.Sprintf("%s%s", w.Prefix, m.Name())
+		} else {
+			name = fmt.Sprintf("%s%s%s%s", w.Prefix, m.Name(), w.MetricSeparator, fieldName)
+		}
+
+		if w.UseRegex {
+			name = sanitizedRegex.ReplaceAllLiteralString(name, "-")
+		} else {
+			name = sanitizedChars.Replace(name)
+		}
+
+		if w.ConvertPaths {
```

Comment:
I'm not so sure about this option being a good idea, it will turn "foo.bar.idle_since" into "foo.bar.idle.since" which IMO is not very nice.

Reply:
wavefront uses '.' delimiters in our metrics browser, and things like measurement: cpu field: usage_idle should become 'cpu.usage.idle' for proper browsing.  This provides a better UX in our product.  This gets especially important in application level metrics where you can return dozens of fields many with the same field prefix.

Reply:
Sometimes the metrics that Telegraf gathers use an underscore only because the name has a space in it, they don't always indicate a hierarchical relationship.

Reply:
We understand that, but Wavefront has made the conscience decision that hierarchies will appear more often than spaces/special chars where an underscore will appear.  We had a few meetings about this, took samples from various input plugins, and decided that converting to dots (default) is the more likely scenario for our customers to have the best possible experience with the product.